### DISCUSSION TITLE
Calling a com object method utilizing an `out` parameter

### DISCUSSION THREAD

**Discussion by msftrncs** [PowerShell/PowerShell#17087] [Q&A]
I am trying to call a method of a com object from an API (Solidworks PDM API) and the method requires using an `out` parameter.

This is the signature that PowerShell/PSReadLine returns during completion of the method:
```
IEdmFile5 GetFileFromPath (string bsFilePath, IEdmFolder5 ppoRetParentFolder)
```

Note it is missing the `out` reference.  Some completions show the requirement of an `out` and others do not and I do not know why.  It doesn't matter if its there or not, the same issue occurs.

However any attempt to call this method returns an error from the either the method or the com object system:
```
OperationStopped: Object reference not set to an instance of an object.
```

If I create a custom type in C# implementing a static function to call this method, all works okay.
``` c#
public class EdmVault5
{
    // helper class to access methods in IEdmVault5 interface that pass objects by reference (out)
    public static IEdmFile5 GetFileFromPath(object vault, string bsFilePath, out IEdmFolder5 ppoRetParentFolder)
    {
        IEdmFolder5 lRetParentFolder;
        IEdmFile5 file = ((IEdmVault5)vault).GetFileFromPath(bsFilePath, out lRetParentFolder);
        ppoRetParentFolder = lRetParentFolder;
        return file;
    }
}
```

Is the issue in PowerShell/.Net/COM Object layer/Interop/API Implementation?   Is there another way to call this method without the helper static method?   The helper method detracts from the flow of scripting since it requires referring to the static method and passing in the source object the method is contained on.

``` Powershell
$file = [SwPdmTools.EdmVault5]::GetFileFromPath($vault,'c:\Vault\SomeFolder\Some.file',[ref]$folder)

# vs

$file = $vault.GetFileFromPath('c:\Vault\SomeFolder\Some.file',[ref]$folder)
```

For reference, the Solidworks PDM API information can be found:
https://help.solidworks.com/2021/English/api/epdmapi/EPDM.Interop.epdm~EPDM.Interop.epdm.IEdmVault5~GetFileFromPath.html

**Comment by msftrncs** [comment]
I have discovered that if `$folder` is already of the returned output type, then the method call will work without errors.  And by output type, I mean already containing the COM object `IEdmFolder5` interface to an existing folder, which I can only do by calling some other method that returns such an object as I cannot do this another way.  PowerShell does not think that `[IEdmFolder5]` and the COM object representing a `IEdmFolder5` are the same kind of object.

**Reply by Empers100** [reply]
Why doesn't Power Shell 7 support the Get-ChildItem cmdlet? PowerShell ISE - supports, but Power Shell 7 - stopped?

**Reply by doctordns** [reply]
Not sure what you mean. Get-ChildItem is very much supported.

**Reply by Empers100** [reply]
![1](https://user-images.githubusercontent.com/96514550/181698435-86ef826d-d963-4277-8ef2-580f70a2107e.jpg)

**Comment by radianrad** [comment]
Did you find a solution? I can create a SW application object, but none of the methods work.

```
PS > $SolidWorks = New-Object -ComObject Sldworks.Application
PS > $SolidWorks | Get-Member

   TypeName: System.__ComObject

Name                      MemberType Definition
----                      ---------- ----------
Equals                    Method     bool Equals(System.Object obj)
GetHashCode               Method     int GetHashCode()
GetLifetimeService        Method     System.Object GetLifetimeService()
GetType                   Method     type GetType()
InitializeLifetimeService Method     System.Object InitializeLifetimeService()
ToString                  Method     string ToString()

```
